hera_mc/cm_health.py

```diff
@@ -0,0 +1,133 @@
+
+    def check_for_duplicate_connections(self):
+        if self.conndict is None:
+            self.get_conndict()
+        if len(self.multiples) == 0:
+            print("No duplications found.")
+            return 0
+        duplicates = []
+        for k in self.multiples:
+            for i, t in enumerate(self.conndict[k]):
+                for j in range(i):
+                    start_i = cm_utils.get_astropytime(self.conndict[k][i][0])
+                    stop_i = cm_utils.get_stopdate(self.conndict[k][i][1])
+                    start_j = cm_utils.get_astropytime(self.conndict[k][j][0])
+                    stop_j = cm_utils.get_stopdate(self.conndict[k][j][1])
+                    print("CM_H:  ", k)
+                    if start_j <= start_i:
+                        if stop_j > start_i:
+                            duplicates.append([k, i, j])
+                    else:
+                        if start_j <= stop_i:
+                            duplicates.append([k, i, j])
+        if len(duplicates):
+            print('{} duplications found.'.format(len(duplicates)))
+            for d in duplicates:
+                k = d[0]
+                start_i = cm_utils.get_astropytime(self.conndict[k][i][0])
+                stop_i = cm_utils.get_astropytime(self.conndict[k][i][1])
+                start_j = cm_utils.get_astropytime(self.conndict[k][j][0])
+                stop_j = cm_utils.get_astropytime(self.conndict[k][j][1])
+                print('\t{} <1>{}-{}  <2>{}-{}'.format(k, start_i, stop_i, start_j, stop_j))
+        else:
+            print('No duplications found.')
+        print('Out of {} connections checked.'.format(self.num_connections))
+        return len(duplicates)
+
+    def check_for_existing_connection(self, connection, at_date='now'):
+        at_date = cm_utils.get_astropytime(at_date)
+        connection = [x.lower() for x in connection]
+        k = ':'.join(connection)
+        if self.conndict is None:
+            self.get_conndict()
+        if k not in self.conndict.keys():
+            return False
+        for t in self.conndict[k]:
+            start_t = cm_utils.get_astropytime(t[0])
+            stop_t = cm_utils.get_astropytime(t[1])
+            if cm_utils.is_active(at_date, start_t, stop_t):
+                print('Duplicate found for connection {} at {}:  {} - {}'.format(k, at_date, start_t, stop_t))
+                return True
+        return False
+
+
+def check_part_for_overlapping_revisions(hpn, session=None):
+    """
+    Checks hpn for parts that overlap in time.  Should be none.
+
+    Returns list of overlapping part revisions
+
+    Parameters:
+    ------------
+    hpn:  hera part name
+    """
+
+    overlap = []
+    revisions = cm_revisions.get_all_revisions(hpn, session)
+    for i, r1 in enumerate(revisions):
+        for j, r2 in enumerate(revisions):
+            if i >= j:
+                continue
```

Comment:
This can be simplified by writing `for j, r2 in enumerate(revisions[:i]):`, I think. But also I don't think the inner index is needed at all; see below.